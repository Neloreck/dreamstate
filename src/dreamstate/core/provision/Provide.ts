import { default as hoistNonReactStatics } from "hoist-non-react-statics";
import { ComponentType } from "react";

import { createManagersObserver } from "@/dreamstate/core/observing/createManagersObserver";
import { createClassWrapDecorator } from "@/dreamstate/polyfills/createClassWrapDecorator";
import { TDreamstateService } from "@/dreamstate/types";

/**
 * Decorator factory.
 * Provide context from context managers.
 * Observes changes and uses default react Providers for data flow.
 */
export function Provide<T extends ComponentType>(
  sources: Array<TDreamstateService<any>>
): ClassDecorator {
  return createClassWrapDecorator(function(targetClass: T): T {
    return hoistNonReactStatics(createManagersObserver(targetClass, sources), targetClass);
  });
}
