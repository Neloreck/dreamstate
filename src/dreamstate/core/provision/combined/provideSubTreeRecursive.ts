import { createElement, ReactNode } from "react";

import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Recursively renders a tree of context providers and wraps the provided component with them.
 *
 * This function generates a React subtree by recursively creating provider components for each context manager
 * in the `sources` array. Each provider in the tree passes down its data to the children components, ensuring that
 * the component tree is properly provided with the necessary context values.
 *
 * @param {ReactNode} bottom - The component (children) that will be wrapped by the context providers.
 * @param {Array<TAnyContextManagerConstructor>} sources - An array of context manager class references that will
 *   provide data via context providers in the React tree.
 * @param {Map<TAnyContextManagerConstructor, TAnyObject>} contextStateRegistry - A map that holds the current
 *   state of each context manager, which will be provided to the corresponding context provider.
 * @param {number} current - The current iteration index for the recursive provider creation process.
 * @returns {ReactNode} A React node representing the tree of context providers, wrapped around the `bottom` component.
 */
export function provideSubTreeRecursive(
  bottom: ReactNode = null,
  sources: Array<TAnyContextManagerConstructor>,
  contextStateRegistry: Map<TAnyContextManagerConstructor, TAnyObject>,
  current: number = 0
): ReactNode {
  /*
   * Wrap children node with providers until all source provider elements are created.
   * Get latest state for provision from current scope registry.
   */
  return current >= sources.length
    ? bottom
    : createElement(
      sources[current].REACT_CONTEXT.Provider,
      { value: contextStateRegistry.get(sources[current]) },
      provideSubTreeRecursive(bottom, sources, contextStateRegistry, current + 1)
    );
}
