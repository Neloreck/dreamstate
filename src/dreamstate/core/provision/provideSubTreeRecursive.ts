import { createElement, ReactElement } from "react";

import { CONTEXT_STATES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function provideSubTreeRecursive(
  bottom: ReactElement,
  sources: Array<TAnyContextManagerConstructor>,
  current: number
): ReactElement {
  return current >= sources.length
    ? bottom
    : createElement(
      sources[current].REACT_CONTEXT.Provider,
      { value: CONTEXT_STATES_REGISTRY.get(sources[current]) },
      provideSubTreeRecursive(bottom, sources, current + 1)
    );
}
