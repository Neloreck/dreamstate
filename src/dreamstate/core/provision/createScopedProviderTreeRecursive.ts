import { createElement, ReactNode } from "react";

import { ScopedProvider } from "@/dreamstate/core/provision/ScopedProvider";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function createScopedProviderTreeRecursive<T extends TAnyObject>(
  bottom: ReactNode = null,
  sources: Array<TAnyContextManagerConstructor>,
  current: number,
  initialState?: T
): ReactNode {
  return current >= sources.length
    ? bottom
    : createElement(
      ScopedProvider,
      { manager: sources[current], initialState },
      createScopedProviderTreeRecursive(bottom, sources, current + 1, initialState)
    );
}
