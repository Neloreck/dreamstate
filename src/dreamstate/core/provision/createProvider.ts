import { FunctionComponent, ReactNode, useCallback, useState } from "react";

import { EMPTY_ARR } from "@/dreamstate/core/internals";
import { useHotObservers } from "@/dreamstate/core/observing/useHotObservers";
import { useStaticObservers } from "@/dreamstate/core/observing/useStaticObservers";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { ContextService } from "@/dreamstate/core/services/ContextService";
import {
  TAnyContextManagerConstructor,
  TAnyContextServiceConstructor,
  TAnyObject
} from "@/dreamstate/types";

export interface IProviderProps<T> {
  initialState?: T;
  partialHotReplacement?: boolean;
  children?: ReactNode;
}

/**
 * Create component for manual provision without HOC/Decorator-like api.
 * Useful if your root is functional component or you are using createComponent api without JSX.
 */
export function createProvider<
  T extends TAnyObject = TAnyObject
>(
  sources: Array<TAnyContextServiceConstructor>
): FunctionComponent<IProviderProps<T>> {
  if (!Array.isArray(sources)) {
    throw new TypeError(
      "Wrong providers parameter supplied. Only array of context services is acceptable."
    );
  }

  for (let it = 0; it < sources.length; it ++) {
    if (!sources[it] || !(sources[it].prototype instanceof ContextService)) {
      throw new TypeError("Only classes extending ContextService can be supplied for provision.");
    }
  }

  /**
   * Check only managers with required provision.
   * Do not include services for subTree rendering but add registering logic for services.
   */
  const managers: Array<TAnyContextManagerConstructor> = sources.filter(function(
    Service: TAnyContextServiceConstructor
  ) {
    return Service.prototype instanceof ContextManager;
  }) as Array<TAnyContextManagerConstructor>;

  /**
   * Create observer component that will handle observing.
   */
  function Observer(props: IProviderProps<T>): ReactNode {
    const [ , forceRender ] = useState({});
    const updateProviders = useCallback(function() {
      forceRender({});
    }, EMPTY_ARR);

    if (props.partialHotReplacement) {
      useHotObservers(sources, props.initialState, updateProviders);
    } else {
      useStaticObservers(sources, props.initialState, updateProviders);
    }

    return provideSubTreeRecursive(props.children, managers, 0);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextServiceConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
