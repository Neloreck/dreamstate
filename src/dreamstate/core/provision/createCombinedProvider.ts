import { FunctionComponent, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { useSourceObserving } from "@/dreamstate/core/provision/useSourceObserving";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Create provider that unifies all data sources and tries to re-render providers on every update.
 */
export function createCombinedProvider<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Observer(props: IProviderProps<T>): ReactNode {
    const registry: Map<TAnyContextManagerConstructor, TAnyObject> = useSourceObserving(sources, props);

    return provideSubTreeRecursive(props.children, sources, registry);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextManagerConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
