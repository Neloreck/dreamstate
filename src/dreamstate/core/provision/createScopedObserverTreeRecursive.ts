import { createElement, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { ScopedObserver } from "@/dreamstate/core/provision/ScopedObserver";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function createScopedObserverTreeRecursive<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>,
  props: IProviderProps<T>,
  current: number = 0
): ReactNode {
  return current >= sources.length
    ? props.children
    : createElement(
      ScopedObserver,
      { manager: sources[current], initialState: props.initialState, dependencies: sources },
      createScopedObserverTreeRecursive(sources, props, current + 1)
    );
}
