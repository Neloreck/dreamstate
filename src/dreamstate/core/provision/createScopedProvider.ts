import { FunctionComponent, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { createScopedProviderTreeRecursive } from "@/dreamstate/core/provision/createScopedProviderTreeRecursive";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

export function createScopedProvider<T extends IProviderProps<any>>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Observer(props: T): ReactNode {
    return createScopedProviderTreeRecursive(props.children, sources, 0, props.initialState);
  }

  Observer.displayName = "DS.Provider";

  return Observer as FunctionComponent<IProviderProps<T>>;
}
