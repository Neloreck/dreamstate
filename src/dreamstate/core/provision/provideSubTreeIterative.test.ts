import { shallow } from "enzyme";
import { createElement } from "react";

import { provideSubTreeIterative } from "@/dreamstate/core/provision/provideSubTreeIterative";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { registerService } from "@/dreamstate/test-utils/registry/registerService";
import { unRegisterService } from "@/dreamstate/test-utils/registry/unRegisterService";
import { NestedContextManager, TestContextManager } from "@/fixtures";

describe("provideSubTreeIterative rendering", () => {
  beforeAll(() => {
    registerService(TestContextManager);
    registerService(NestedContextManager);
  });

  afterAll(() => {
    unRegisterService(TestContextManager);
    unRegisterService(NestedContextManager);
  });

  const provide = (method: typeof provideSubTreeIterative | typeof provideSubTreeRecursive) => {
    shallow(
      createElement(
        "div",
        {},
        method(
          createElement("div", {}, "bottom"),
          [
            TestContextManager,
            NestedContextManager,
            TestContextManager
          ],
          0
        )
      )
    );
  };

  it("Should match snapshot", () => {
    const iterative = provide(provideSubTreeIterative);

    expect(iterative).toMatchSnapshot();
  });

  it("Should be same as iterative implementation", () => {
    const iterative = provide(provideSubTreeIterative);
    const recursive = provide(provideSubTreeRecursive);

    expect(recursive).toEqual(iterative);
  });
});
