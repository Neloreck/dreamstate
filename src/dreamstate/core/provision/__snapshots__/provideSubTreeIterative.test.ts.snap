// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provideSubTreeIterative rendering. Should match snapshot. 1`] = `undefined`;
